File: LeetCode题目集/205. Isomorphic Strings(easy).java
import java.util.Hashtable;
import java.util.Map;

/*
 * @Author: SourDumplings
 * @Date: 2019-08-18 10:00:02
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/isomorphic-strings/
 */

class Solution
{
    public boolean isIsomorphic(String s, String t)
    {
        int l = s.length();
        if (l != t.length())
        {
            return false;
        }
        Map<Character, Character> m1 = new Hashtable<>();
        Map<Character, Character> m2 = new Hashtable<>();
        for (int i = 0; i < l; i++)
        {
            Character c1 = s.charAt(i);
            Character c2 = t.charAt(i);
            if (m1.containsKey(c1))
            {
                if (m1.get(c1) != c2)
                {
                    return false;
                }
            }
            else
            {
                if (m2.containsKey(c2))
                {
                    return false;
                }
                m1.put(c1, c2);
                m2.put(c2, c1);
            }
        }
        return true;
    }
}

File: OJ practices/LeetCode题目集/205. Isomorphic Strings(easy).cpp
/**
 * @file 205. Isomorphic Strings(easy).cpp
 * @author SourDumplings (changzheng300@foxmail.com)
 * @brief 
 * @version 1.0.0
 * @date 2024-04-13
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 * https://leetcode.cn/problems/isomorphic-strings/description/?envType=study-plan-v2&envId=top-interview-150
 */

class Solution
{
public:
    bool isIsomorphic(string s, string t)
    {
        int l = s.length();
        char m[200];
        char rm[200];
        memset(m, -1, sizeof(m));
        memset(rm, -1, sizeof(rm));
        for (int i = 0; i < l; ++i)
        {
            if (m[s[i]] == -1)
            {
                m[s[i]] = t[i];
                if (rm[t[i]] != -1)
                {
                    return false;
                }
                rm[t[i]] = s[i];
                
            }
            else if (m[s[i]] != t[i])
            {
                return false;
            }
            else if (rm[t[i]] != s[i])
            {
                return false;
            }
            
        }
        return true;
    }
};